module/zfs/zio_crypt.c

```diff
@@ -0,0 +1,1373 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License, Version 1.0 only
+ * (the "License").  You may not use this file except in compliance
+ * with the License.
+ *
+ * You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
+ * or http://www.opensolaris.org/os/licensing.
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+/*
+ * Copyright (c) 2016, Datto, Inc. All rights reserved.
+ */
+
+#include <sys/zio_crypt.h>
+#include <sys/dmu.h>
+#include <sys/dnode.h>
+#include <sys/fs/zfs.h>
+#include <sys/zio.h>
+#include <sys/zil.h>
+
+/*
+ * This file is responsible for handling all of the details of generating
+ * encryption parameters and performing encryption.
+ *
+ * BLOCK ENCRYPTION PARAMETERS:
+ * Encryption Algorithm (crypt):
+ * The encryption algorithm and mode we are going to use. We currently support
+ * AES-GCM and AES-CCM in 128, 192, and 256 bits. All encryption parameters are
+ * stored in little endian format (regardless of the host machine's byteorder).
+ *
+ * Plaintext:
+ * The unencrypted data that we want to encrypt
+ *
+ * Initialization Vector (IV):
+ * An initialization vector for the encryption algorithms. This is
+ * used to "tweak" the encryption algorithms so that equivalent blocks of
+ * data are encrypted into different ciphertext outputs. Different modes
+ * of encryption have different requirements for the IV. AES-GCM and AES-CCM
+ * require that an IV is never reused with the same encryption key. This
+ * value is stored unencrypted and must simply be provided to the decryption
+ * function. We use a 96 bit IV (as recommended by NIST). For non-dedup blocks
+ * we derive the IV randomly. The first 64 bits of the IV are stored in the
+ * second word of DVA[2] and the remaining 32 bits are stored in the upper 32
+ * bits of blk_fill. For most object types this is safe because we only encrypt
+ * level 0 blocks which means that the fill count will be 1. For DMU_OT_DNODE
+ * blocks the fill count is instead used to indicate the number of free dnodes
```

Comment:
Interesting, that seems good for `dsl_dataset_space(usedobjsp)` but potentially problematic for `dmu_object_alloc_dnsize()`, because when it looks for a L1 that's at most 1/4 full, it's measuring that fullness by number of dnodes rather than number of slots.  For example, if every dnode uses 4 slots, then no block will be more than 1/4 "full".  @nedbass is that right?